sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/SparkInternalsBridge.scala

```diff
@@ -0,0 +1,82 @@
+import com.azure.cosmos.implementation.guava25.base.Strings.emptyToNull
+import com.azure.cosmos.spark.diagnostics.BasicLoggingTrait
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.util.AccumulatorV2
+
+import java.util.Locale
+import java.util.concurrent.atomic.AtomicBoolean
+
+object SparkInternalsBridge extends BasicLoggingTrait {
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_PROPERTY = "COSMOS.SPARK_REFLECTION_ACCESS_ALLOWED"
+  private val SPARK_REFLECTION_ACCESS_ALLOWED_VARIABLE = "COSMOS_SPARK_REFLECTION_ACCESS_ALLOWED"
+
+  private val DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED = true
+
+  def getSparkReflectionAccessAllowed: Boolean = {
+    val allowedText = System.getProperty(
+      SPARK_REFLECTION_ACCESS_ALLOWED_PROPERTY,
+      firstNonNull(
+        emptyToNull(System.getenv.get(SPARK_REFLECTION_ACCESS_ALLOWED_VARIABLE)),
+        String.valueOf(DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED)))
+
+    try {
+      java.lang.Boolean.valueOf(allowedText.toUpperCase(Locale.ROOT))
+    }
+    catch {
+      case e: Exception =>
+        logError(s"Parsing spark reflection access allowed ${allowedText} failed. Using the default ${DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED}.", e)
+        DEFAULT_SPARK_REFLECTION_ACCESS_ALLOWED
+    }
+  }
+
+  private final lazy val reflectionAccessAllowed = new AtomicBoolean(getSparkReflectionAccessAllowed)
+
+  def getInternalCustomTaskMetricsAsSQLMetric(knownCosmosMetricNames: Set[String]) : Map[String, SQLMetric] = {
+
+    if (!reflectionAccessAllowed.get) {
+      Map.empty[String, SQLMetric]
+    } else {
+      Option.apply(TaskContext.get()) match {
+        case Some(taskCtx) => getInternalCustomTaskMetricsAsSQLMetricInternal(knownCosmosMetricNames, taskCtx)
+        case None => Map.empty[String, SQLMetric]
+      }
+    }
+  }
+
+  private def getAccumulators(taskCtx: TaskContext): Seq[AccumulatorV2[_, _]] = {
+    val taskMetrics:Object = taskCtx.taskMetrics()
+    taskMetrics.getClass.getMethod("accumulators").invoke(taskMetrics).asInstanceOf[Seq[AccumulatorV2[_, _]]]
+  }
+
+  private def getInternalCustomTaskMetricsAsSQLMetricInternal(
+    knownCosmosMetricNames: Set[String],
+    taskCtx: TaskContext): Map[String, SQLMetric] = {
+
+    try {
+      val accumulators: Seq[AccumulatorV2[_, _]] = getAccumulators(taskCtx)
+      accumulators
+        .filter(accumulable => accumulable.isInstanceOf[SQLMetric]
+          && accumulable.name.isDefined
+          && knownCosmosMetricNames.contains(accumulable.name.get))
+        .map(accumulable => {
+          val sqlMetric = accumulable.asInstanceOf[SQLMetric]
+          sqlMetric.name.get -> sqlMetric
+        })
+        .toMap[String, SQLMetric]
+    } catch {
+      case e: Exception => {
+        logInfo(s"Could not populate custom metrics - Error ${e.getMessage}", e)
+
+        // reflection failed - disabling it for the future
+        reflectionAccessAllowed.set(false)
```

Comment:
the exception being captured here is generic errors, should we only disallow for `InvocationTargetException` or similar?

Reply:
Keep ctaching generic exception intentionally - but scoped the try-block to the relevant reflection code.